// @sentry/opentelemetry-node/cjs/utils/captureExceptionForTimedEvent.js
Object.defineProperty(exports, '__esModule', { value: true });

const semanticConventions = require('@opentelemetry/semantic-conventions');
const utils = require('@sentry/utils');

/**
 * Maybe capture a Sentry exception for an OTEL timed event.
 * This will check if the event is exception-like and in that case capture it as an exception.
 */
// eslint-disable-next-line deprecation/deprecation
function maybeCaptureExceptionForTimedEvent(hub, event, otelSpan) {
  if (event.name !== 'exception') {
    return;
  }

  const attributes = event.attributes;
  if (!attributes) {
    return;
  }

  const message = attributes[semanticConventions.SemanticAttributes.EXCEPTION_MESSAGE];

  if (typeof message !== 'string') {
    return;
  }

  const syntheticError = new Error(message);

  const stack = attributes[semanticConventions.SemanticAttributes.EXCEPTION_STACKTRACE];
  if (utils.isString(stack)) {
    syntheticError.stack = stack;
  }

  const type = attributes[semanticConventions.SemanticAttributes.EXCEPTION_TYPE];
  if (utils.isString(type)) {
    syntheticError.name = type;
  }

  // eslint-disable-next-line deprecation/deprecation
  hub.captureException(syntheticError, {
    captureContext: otelSpan
      ? {
          contexts: {
            otel: {
              attributes: otelSpan.attributes,
              resource: otelSpan.resource.attributes,
            },
            trace: {
              trace_id: otelSpan.spanContext().traceId,
              span_id: otelSpan.spanContext().spanId,
              parent_span_id: otelSpan.parentSpanId,
            },
          },
        }
      : undefined,
  });
}

exports.maybeCaptureExceptionForTimedEvent = maybeCaptureExceptionForTimedEvent;
